### PR TITLE
Add ability to parse hours/months without leading zeros

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Formatters `%l`/`%k`/`%c` in function `parseDateTime()` are now able to parse hours/months without leading zeros, e.g. `select parseDateTime('2023-11-26 8:14', '%F %k:%i')` now works. Set `parsedatetime_parse_without_leading_zeros = 0` to restore the previous behavior which required two digits. Function `formatDateTime` is now also able to print hours/months without leading zeros. This is controlled by setting `formatdatetime_format_without_leading_zeros` but off by default to not break existing use cases.

### TL;DR;

Recently I was looking into one dataset, that has such format.

So I've changed the meaning of the following formatters for
formatDateTime()/parseDateTime(), like in other implementations [1]:

- %l/%k - to handle hours without leading zeros
- %c - handle months without leading zeros

  [1]: https://www.php.net/strftime

And since this is kind of behaviour change (even though it is unlikely
will break something), there is a new setting to control this new
behaviour:
- `parsedatetime_parse_without_leading_zeros=1` (**default ON**)
- `formatdatetime_format_without_leading_zeros=0` (**default OFF**, since this could change users output)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit dfadfffe18b8c8aa4ebf4f10b0f5289968755797 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55872/dfadfffe18b8c8aa4ebf4f10b0f5289968755797/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6588594411">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55872/dfadfffe18b8c8aa4ebf4f10b0f5289968755797/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by azat** [comment]
Can someone take a look?

File: docs/en/sql-reference/functions/date-time-functions.md [hunk 1]
```diff
@@ -2492,8 +2492,8 @@ Using replacement fields, you can define a pattern for the resulting string. “
 | %i       | minute (00-59)                                          | 33         |
 | %I       | hour in 12h format (01-12)                              | 10         |
 | %j       | day of the year (001-366)                               | 002        |
-| %k       | hour in 24h format (00-23)                              | 22         |
-| %l       | hour in 12h format (01-12)                              | 09         |
+| %k       | hour in 24h format (00-23)                              | 2          |
```
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/date-time-functions.md:2495]: Then it should be `0-23`

**Review by alexey-milovidov — APPROVED** [review]
The idea LGTM, only one change remaining.

File: docs/en/sql-reference/functions/date-time-functions.md (hunk 1 above)
Comment (azat) [reply docs/en/sql-reference/functions/date-time-functions.md:2495]: Indeed, thanks, adjusted.

File: src/Functions/parseDateTime.cpp [hunk 2]
```diff
@@ -1123,6 +1123,14 @@ namespace
                 return cur;
             }
 
+            static Pos mysqlHour12NoLeadingZero(Pos cur, Pos end, const String & fragment, DateTime & date)
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:1126]: Cosmetic: To stay consistent with the terminology used within the rest of the file: `mysqlHour12SpacePadded`.

Similarly, l. 1142: `mysqlHour24SpacePadded`

Comment (rschu1ze) [reply src/Functions/parseDateTime.cpp:1126]: Cosmetic: Let's reorder l. 1126-1132 after l. 1134-1140. (same for the 24 h variant)

File: tests/queries/0_stateless/02902_parseDateTime_hour_no_leading_zero.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
+-- { echo }
+
+-- %h
+select parseDateTime('Aug 13, 2022, 7:58:32 PM', '%b %e, %G, %h:%i:%s %p', 'UTC'); -- { serverError CANNOT_PARSE_DATETIME }
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02902_parseDateTime_hour_no_leading_zero.sql:4]: Let's adjust/extend the existing test 02668_parse_datetime.sql instead.

File: docs/en/sql-reference/functions/date-time-functions.md [hunk 4]
```diff
@@ -2553,8 +2553,8 @@ Using replacement fields, you can define a pattern for the resulting string. “
 | %i       | minute (00-59)                                          | 33         |
 | %I       | hour in 12h format (01-12)                              | 10         |
 | %j       | day of the year (001-366)                               | 002        |
-| %k       | hour in 24h format (00-23)                              | 22         |
-| %l       | hour in 12h format (01-12)                              | 09         |
+| %k       | hour in 24h format (0-23)                               | 2          |
```
Comment (rschu1ze) [inline docs/en/sql-reference/functions/date-time-functions.md:2556]: There are really two functions involved here: `parseDateTime()` and `formatDateTime()` and both functions "mirror" each other.

About `parseDateTime()`: `%l`/`%k` now accept single-digit hours. This is an incompatible change (as stated by the changelog entry) in a quite popular function. It would be better if there was a setting that restores the legacy. There are two reasons:
1. I know of situations in the past where incompatible output in functions broke existing use cases (queries, table/partition definitions). If this happens (and it does not really matter how unlikely we thing this is - there are always "weird misuses" of ClickHouse out there), a setting comes in handy.
2. The new behavior increases compatibility with other DBMS (good) but it also increases ambiguity in certain edge cases such as `SELECT parseDateTime('112', '%k%k');` - will the result have `2` or `12` as the hour component? Note that this can't happen with fixed-width formatters and we should give users a possibility to switch back.

About the name of the setting: As far as I see, `%c` (month) generally has the same problem: 2 digits in ClickHouse, 1-2 digits in MySQL's `FORMAT_DATE()` ([docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format)) function (after which ClickHouse's `formatDateTIme()` is modeled). `%c` is not addressed by this PR - if you like to, it would be very appreciated. So what about: `formatdatetime_lkc_is_space_padded` (default: `true`)? There could be a "Note 3" in this document, see l. 2578ff.

Then there is `formatDateTime()`: users expect that `parseDateTime(formatDateTime(...))` is idempotent, so similar change would happen there. `formatDateTime()` has [an optimization](https://github.com/ClickHouse/ClickHouse/blob/8c253b9e3e46c416623047d2df876eab0e2eb5ba/src/Functions/formatDateTime.cpp#L873) which will unfortunately no longer work for `%l`/`%k` (`%c`) but it is what it is. Ideally, the new setting is also respected in `formatDateTime()`.

File: src/Functions/parseDateTime.cpp (hunk 2 above)
Comment (azat) [reply src/Functions/parseDateTime.cpp:1126]: Ok, adjusted.

File: tests/queries/0_stateless/02902_parseDateTime_hour_no_leading_zero.sql (hunk 3 above)
Comment (azat) [reply tests/queries/0_stateless/02902_parseDateTime_hour_no_leading_zero.sql:4]: Ok, updated. (AFAIR the policy was different).

File: docs/en/sql-reference/functions/date-time-functions.md (hunk 4 above)
Comment (azat) [reply docs/en/sql-reference/functions/date-time-functions.md:2556]: >1-2 digits in MySQL's FORMAT_DATE() ([docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format)) function (after which ClickHouse's formatDateTIme() is modeled)

Actually MySQL accept any variant

```sql
mysql> select version();
+-----------+
| version() |
+-----------+
| 8.2.0     |
+-----------+

mysql> SELECT STR_TO_DATE('1717:1', '%H');
+-----------------------------+
| STR_TO_DATE('1717:1', '%H') |
+-----------------------------+
| 17:00:00                    |
+-----------------------------+

mysql> SELECT STR_TO_DATE('7', '%H');
+------------------------+
| STR_TO_DATE('7', '%H') |
+------------------------+
| 07:00:00               |
+------------------------+
1 row in set (0.00 sec)

mysql> SELECT STR_TO_DATE('7', '%k');
+------------------------+
| STR_TO_DATE('7', '%k') |
+------------------------+
| 07:00:00               |
+------------------------+
1 row in set (0.00 sec)
```

Comment (azat) [reply docs/en/sql-reference/functions/date-time-functions.md:2556]: >So what about: formatdatetime_lkc_is_space_padded (default: true)

Looks too specific, I will go with `formatdatetime_parsedatetime_enable_format_with_leading_zeros`

File: src/Functions/parseDateTime.cpp (hunk 2 above)
Comment (azat) [reply src/Functions/parseDateTime.cpp:1126]: BTW why it should have `SpacePadded`, the space padded version is indeed padded with space, I will revert to `WithoutLeadingZero`

**Comment by alexey-milovidov** [comment]
@rschu1ze Is it true that these formatting modifiers were added recently, we did it only for MySQL compatibility, and they are generally weird? So, people get used to regular modifiers with double digits, that were implemented a long time ago, and we just recently added this as an excuse for MySQL? Then we can change the behavior even without introducing a setting.

**Comment by alexey-milovidov** [comment]
In contrast to `formatDateTime`, `parseDateTime` can afford to parse all variants (two digits, one digit padded with whitespace, one digit not padded) with both of the modifiers.

File: docs/en/sql-reference/functions/date-time-functions.md (hunk 4 above)
Comment (azat) [reply docs/en/sql-reference/functions/date-time-functions.md]: So:
- I've added support of month without leading zeros (`%c`)
- add setting to control behaviour of `parseDateTime` - `parsedatetime_enable_format_without_leading_zeros=1` (ON by default)
- adjust `formatDateTime` to this format changes as well, by only under `formatdatetime_enable_format_without_leading_zeros=1`, it is OFF by default (since this can break some user output, i.e. exports)
- adjusted settings changes
- updated the description (and of course commit message as well) to reflect this changes

**Comment by azat** [comment]
Now there is a setting for "everything"...

**Review by alexey-milovidov — APPROVED** [review]
Neat.

**Comment by azat** [comment]
>PullRequestCI / FunctionalStatelessTestReleaseAnalyzer / Stateless tests (release, analyzer) (pull_request) Failing after 360m

Looks like internal CI issue

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
> @rschu1ze Is it true that these formatting modifiers were added recently, we did it only for MySQL compatibility, and they are generally weird? So, people get used to regular modifiers with double digits, that were implemented a long time ago, and we just recently added this as an excuse for MySQL? Then we can change the behavior even without introducing a setting.

%c/%l/%k were not added recently for MySQL compatibility, they exist since parse/formatDateTime() exist. And no, they are not generally weird. The use case for settings is not to make the behavior more complex for users but to have a safety net in case something breaks unexpectedly. I am good with removing the settings after 2 or 3 LTS releases.

**Comment by rschu1ze** [comment]
ClickHouse Integration Tests (tsan) [4/6]
- test_system_clusters_actual_information/test.py::test: #55168
